File: src/brs/http/ParameterParser.java
package brs.http;

import brs.Constants;
import brs.crypto.EncryptedData;
import brs.http.common.Parameters;
import brs.util.Convert;

import javax.servlet.http.HttpServletRequest;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;

import static brs.http.JSONResponses.*;
import static brs.http.common.Parameters.*;

final class ParameterParser {

  static long getFeeNQT(HttpServletRequest req) throws ParameterException {
    String feeValueNQT = Convert.INSTANCE.emptyToNull(req.getParameter(FEE_NQT_PARAMETER));
    if (feeValueNQT == null) {
      throw new ParameterException(MISSING_FEE);
    }
    long feeNQT;
    try {
      feeNQT = Long.parseLong(feeValueNQT);
    } catch (RuntimeException e) {
      throw new ParameterException(INCORRECT_FEE);
    }
    if (feeNQT < 0 || feeNQT >= Constants.MAX_BALANCE_NQT) {
      throw new ParameterException(INCORRECT_FEE);
    }
    return feeNQT;
  }

  static long getPriceNQT(HttpServletRequest req) throws ParameterException {
    String priceValueNQT = Convert.INSTANCE.emptyToNull(req.getParameter(PRICE_NQT_PARAMETER));
    if (priceValueNQT == null) {
      throw new ParameterException(MISSING_PRICE);
    }
    long priceNQT;
    try {
      priceNQT = Long.parseLong(priceValueNQT);
    } catch (RuntimeException e) {
      throw new ParameterException(INCORRECT_PRICE);
    }
    if (priceNQT <= 0 || priceNQT > Constants.MAX_BALANCE_NQT) {
      throw new ParameterException(INCORRECT_PRICE);
    }
    return priceNQT;
  }

  static long getQuantityQNT(HttpServletRequest req) throws ParameterException {
    String quantityValueQNT = Convert.INSTANCE.emptyToNull(req.getParameter(QUANTITY_QNT_PARAMETER));
    if (quantityValueQNT == null) {
      throw new ParameterException(MISSING_QUANTITY);
    }
    long quantityQNT;
    try {
      quantityQNT = Long.parseLong(quantityValueQNT);
    } catch (RuntimeException e) {
      throw new ParameterException(INCORRECT_QUANTITY);
    }
    if (quantityQNT <= 0 || quantityQNT > Constants.MAX_ASSET_QUANTITY_QNT) {
      throw new ParameterException(INCORRECT_ASSET_QUANTITY);
    }
    return quantityQNT;
  }

  static long getOrderId(HttpServletRequest req) throws ParameterException {
    String orderValue = Convert.INSTANCE.emptyToNull(req.getParameter(ORDER_PARAMETER));
    if (orderValue == null) {
      throw new ParameterException(MISSING_ORDER);
    }
    try {
      return Convert.INSTANCE.parseUnsignedLong(orderValue);
    } catch (RuntimeException e) {
      throw new ParameterException(INCORRECT_ORDER);
    }
  }

  static int getGoodsQuantity(HttpServletRequest req) throws ParameterException {
    String quantityString = Convert.INSTANCE.emptyToNull(req.getParameter(QUANTITY_PARAMETER));
    try {
      int quantity = Integer.parseInt(quantityString);
      if (quantity < 0 || quantity > Constants.MAX_DGS_LISTING_QUANTITY) {
        throw new ParameterException(INCORRECT_QUANTITY);
      }
      return quantity;
    } catch (NumberFormatException e) {
      throw new ParameterException(INCORRECT_QUANTITY);
    }
  }

  static EncryptedData getEncryptedGoods(HttpServletRequest req) throws ParameterException {
    String data = Convert.INSTANCE.emptyToNull(req.getParameter(GOODS_DATA_PARAMETER));
    String nonce = Convert.INSTANCE.emptyToNull(req.getParameter(GOODS_NONCE_PARAMETER));
    if (data != null && nonce != null) {
      try {
        return new EncryptedData(Convert.INSTANCE.parseHexString(data), Convert.INSTANCE.parseHexString(nonce));
      } catch (RuntimeException e) {
        throw new ParameterException(INCORRECT_DGS_ENCRYPTED_GOODS);
      }
    }
    return null;
  }

  static String getSecretPhrase(HttpServletRequest req) throws ParameterException {
    String secretPhrase = Convert.INSTANCE.emptyToNull(req.getParameter(SECRET_PHRASE_PARAMETER));
    if (secretPhrase == null) {
      throw new ParameterException(MISSING_SECRET_PHRASE);
    }
    return secretPhrase;
  }

  static int getTimestamp(HttpServletRequest req) throws ParameterException {
    String timestampValue = Convert.INSTANCE.emptyToNull(req.getParameter(TIMESTAMP_PARAMETER));
    if (timestampValue == null) {
      return 0;
    }
    int timestamp;
    try {
      timestamp = Integer.parseInt(timestampValue);
    } catch (NumberFormatException e) {
      throw new ParameterException(INCORRECT_TIMESTAMP);
    }
    if (timestamp < 0) {
      throw new ParameterException(INCORRECT_TIMESTAMP);
    }
    return timestamp;
  }

  static long getRecipientId(HttpServletRequest req) throws ParameterException {
    String recipientValue = Convert.INSTANCE.emptyToNull(req.getParameter(RECIPIENT_PARAMETER));
    if (recipientValue == null || Parameters.isZero(recipientValue)) {
      throw new ParameterException(MISSING_RECIPIENT);
    }
    long recipientId;
    try {
      recipientId = Convert.INSTANCE.parseAccountId(recipientValue);
    } catch (RuntimeException e) {
      throw new ParameterException(INCORRECT_RECIPIENT);
    }
    if (recipientId == 0) {
      throw new ParameterException(INCORRECT_RECIPIENT);
    }
    return recipientId;
  }

  static long getSellerId(HttpServletRequest req) throws ParameterException {
    String sellerIdValue = Convert.INSTANCE.emptyToNull(req.getParameter(SELLER_PARAMETER));
    try {
      return Convert.INSTANCE.parseAccountId(sellerIdValue);
    } catch (RuntimeException e) {
      throw new ParameterException(INCORRECT_RECIPIENT);
    }
  }

  static long getBuyerId(HttpServletRequest req) throws ParameterException {
    String buyerIdValue = Convert.INSTANCE.emptyToNull(req.getParameter(BUYER_PARAMETER));
    try {
      return Convert.INSTANCE.parseAccountId(buyerIdValue);
    } catch (RuntimeException e) {
      throw new ParameterException(INCORRECT_RECIPIENT);
    }
  }

  static int getFirstIndex(HttpServletRequest req) {
    int firstIndex;
    try {
      firstIndex = Integer.parseInt(req.getParameter(FIRST_INDEX_PARAMETER));
      if (firstIndex < 0) {
        return 0;
      }
    } catch (NumberFormatException e) {
      return 0;
    }
    return firstIndex;
  }

  static int getLastIndex(HttpServletRequest req) {
    int lastIndex;
    try {
      lastIndex = Integer.parseInt(req.getParameter(LAST_INDEX_PARAMETER));
      if (lastIndex < 0) {
        return Integer.MAX_VALUE;
      }
    } catch (NumberFormatException e) {
      return Integer.MAX_VALUE;
    }
    return lastIndex;
  }

  private ParameterParser() {
  } // never

  public static byte[] getCreationBytes(HttpServletRequest req) throws ParameterException {
    try {
      return Convert.INSTANCE.parseHexString(req.getParameter(CREATION_BYTES_PARAMETER));
    } catch (RuntimeException e) {
      throw new ParameterException(INCORRECT_CREATION_BYTES);
    }


  }

  public static String getATLong(HttpServletRequest req) {
    String hex = req.getParameter(HEX_STRING_PARAMETER);
    ByteBuffer bf = ByteBuffer.allocate(8);
    bf.order(ByteOrder.LITTLE_ENDIAN);
    bf.put(Convert.INSTANCE.parseHexString(hex));

      return Convert.INSTANCE.toUnsignedLong(bf.getLong(0));
  }

  public static long getAmountNQT(HttpServletRequest req) throws ParameterException {
    String amountValueNQT = Convert.INSTANCE.emptyToNull(req.getParameter(AMOUNT_NQT_PARAMETER));
    if (amountValueNQT == null) {
      throw new ParameterException(MISSING_AMOUNT);
    }
    long amountNQT;
    try {
      amountNQT = Long.parseLong(amountValueNQT);
    } catch (RuntimeException e) {
      throw new ParameterException(INCORRECT_AMOUNT);
    }
    if (amountNQT <= 0 || amountNQT >= Constants.MAX_BALANCE_NQT) {
      throw new ParameterException(INCORRECT_AMOUNT);
    }
    return amountNQT;
  }
}

File: src/main/java/brs/http/ParameterParser.kt
package brs.http

import brs.Constants
import brs.crypto.EncryptedData
import brs.http.JSONResponses.INCORRECT_AMOUNT
import brs.http.JSONResponses.INCORRECT_ASSET_QUANTITY
import brs.http.JSONResponses.INCORRECT_CREATION_BYTES
import brs.http.JSONResponses.INCORRECT_DGS_ENCRYPTED_GOODS
import brs.http.JSONResponses.INCORRECT_FEE
import brs.http.JSONResponses.INCORRECT_ORDER
import brs.http.JSONResponses.INCORRECT_PRICE
import brs.http.JSONResponses.INCORRECT_QUANTITY
import brs.http.JSONResponses.INCORRECT_RECIPIENT
import brs.http.JSONResponses.INCORRECT_TIMESTAMP
import brs.http.JSONResponses.MISSING_AMOUNT
import brs.http.JSONResponses.MISSING_FEE
import brs.http.JSONResponses.MISSING_ORDER
import brs.http.JSONResponses.MISSING_PRICE
import brs.http.JSONResponses.MISSING_QUANTITY
import brs.http.JSONResponses.MISSING_RECIPIENT
import brs.http.JSONResponses.MISSING_SECRET_PHRASE
import brs.http.common.Parameters
import brs.http.common.Parameters.AMOUNT_NQT_PARAMETER
import brs.http.common.Parameters.BUYER_PARAMETER
import brs.http.common.Parameters.CREATION_BYTES_PARAMETER
import brs.http.common.Parameters.FEE_NQT_PARAMETER
import brs.http.common.Parameters.FIRST_INDEX_PARAMETER
import brs.http.common.Parameters.GOODS_DATA_PARAMETER
import brs.http.common.Parameters.GOODS_NONCE_PARAMETER
import brs.http.common.Parameters.HEX_STRING_PARAMETER
import brs.http.common.Parameters.LAST_INDEX_PARAMETER
import brs.http.common.Parameters.ORDER_PARAMETER
import brs.http.common.Parameters.PRICE_NQT_PARAMETER
import brs.http.common.Parameters.QUANTITY_PARAMETER
import brs.http.common.Parameters.QUANTITY_QNT_PARAMETER
import brs.http.common.Parameters.RECIPIENT_PARAMETER
import brs.http.common.Parameters.SECRET_PHRASE_PARAMETER
import brs.http.common.Parameters.SELLER_PARAMETER
import brs.http.common.Parameters.TIMESTAMP_PARAMETER
import brs.util.convert.*
import java.nio.ByteBuffer
import java.nio.ByteOrder
import javax.servlet.http.HttpServletRequest

internal object ParameterParser {

    fun getFeeNQT(request: HttpServletRequest): Long {
        val feeValueNQT = request.getParameter(FEE_NQT_PARAMETER).emptyToNull()
                ?: throw ParameterException(MISSING_FEE)
        val feeNQT: Long
        try {
            feeNQT = java.lang.Long.parseLong(feeValueNQT)
        } catch (e: RuntimeException) {
            throw ParameterException(INCORRECT_FEE)
        }

        if (feeNQT < 0 || feeNQT >= Constants.MAX_BALANCE_NQT) {
            throw ParameterException(INCORRECT_FEE)
        }
        return feeNQT
    }

    fun getPriceNQT(request: HttpServletRequest): Long {
        val priceValueNQT = request.getParameter(PRICE_NQT_PARAMETER).emptyToNull()
                ?: throw ParameterException(MISSING_PRICE)
        val priceNQT: Long
        try {
            priceNQT = java.lang.Long.parseLong(priceValueNQT)
        } catch (e: RuntimeException) {
            throw ParameterException(INCORRECT_PRICE)
        }

        if (priceNQT <= 0 || priceNQT > Constants.MAX_BALANCE_NQT) {
            throw ParameterException(INCORRECT_PRICE)
        }
        return priceNQT
    }

    fun getQuantityQNT(request: HttpServletRequest): Long {
        val quantityValueQNT = request.getParameter(QUANTITY_QNT_PARAMETER).emptyToNull()
                ?: throw ParameterException(MISSING_QUANTITY)
        val quantityQNT: Long
        try {
            quantityQNT = java.lang.Long.parseLong(quantityValueQNT)
        } catch (e: RuntimeException) {
            throw ParameterException(INCORRECT_QUANTITY)
        }

        if (quantityQNT <= 0 || quantityQNT > Constants.MAX_ASSET_QUANTITY_QNT) {
            throw ParameterException(INCORRECT_ASSET_QUANTITY)
        }
        return quantityQNT
    }

    fun getOrderId(request: HttpServletRequest): Long {
        val orderValue = request.getParameter(ORDER_PARAMETER).emptyToNull()
                ?: throw ParameterException(MISSING_ORDER)
        try {
            return orderValue.parseUnsignedLong()
        } catch (e: RuntimeException) {
            throw ParameterException(INCORRECT_ORDER)
        }

    }

    fun getGoodsQuantity(request: HttpServletRequest): Int {
        val quantityString = request.getParameter(QUANTITY_PARAMETER).emptyToNull()
        try {
            val quantity = Integer.parseInt(quantityString!!)
            if (quantity < 0 || quantity > Constants.MAX_DGS_LISTING_QUANTITY) {
                throw ParameterException(INCORRECT_QUANTITY)
            }
            return quantity
        } catch (e: NumberFormatException) {
            throw ParameterException(INCORRECT_QUANTITY)
        }

    }

    fun getEncryptedGoods(request: HttpServletRequest): EncryptedData? {
        val data = request.getParameter(GOODS_DATA_PARAMETER).emptyToNull()
        val nonce = request.getParameter(GOODS_NONCE_PARAMETER).emptyToNull()
        if (data != null && nonce != null) {
            try {
                return EncryptedData(data.parseHexString(), nonce.parseHexString())
            } catch (e: RuntimeException) {
                throw ParameterException(INCORRECT_DGS_ENCRYPTED_GOODS)
            }

        }
        return null
    }

    fun getSecretPhrase(request: HttpServletRequest): String {
        return request.getParameter(SECRET_PHRASE_PARAMETER).emptyToNull()
                ?: throw ParameterException(MISSING_SECRET_PHRASE)
    }

    fun getTimestamp(request: HttpServletRequest): Int {
        val timestampValue = request.getParameter(TIMESTAMP_PARAMETER).emptyToNull() ?: return 0
        val timestamp: Int
        try {
            timestamp = Integer.parseInt(timestampValue)
        } catch (e: NumberFormatException) {
            throw ParameterException(INCORRECT_TIMESTAMP)
        }

        if (timestamp < 0) {
            throw ParameterException(INCORRECT_TIMESTAMP)
        }
        return timestamp
    }

    fun getRecipientId(request: HttpServletRequest): Long {
        val recipientValue = request.getParameter(RECIPIENT_PARAMETER).emptyToNull()
        if (recipientValue == null || Parameters.isZero(recipientValue)) {
            throw ParameterException(MISSING_RECIPIENT)
        }
        val recipientId: Long
        try {
            recipientId = recipientValue.parseAccountId()
        } catch (e: RuntimeException) {
            throw ParameterException(INCORRECT_RECIPIENT)
        }

        if (recipientId == 0L) {
            throw ParameterException(INCORRECT_RECIPIENT)
        }
        return recipientId
    }

    fun getSellerId(request: HttpServletRequest): Long {
        val sellerIdValue = request.getParameter(SELLER_PARAMETER).emptyToNull()
        try {
            return sellerIdValue!!.parseAccountId()
        } catch (e: RuntimeException) {
            throw ParameterException(INCORRECT_RECIPIENT)
        }

    }

    fun getBuyerId(request: HttpServletRequest): Long {
        val buyerIdValue = request.getParameter(BUYER_PARAMETER).emptyToNull()
        try {
            return buyerIdValue!!.parseAccountId()
        } catch (e: RuntimeException) {
            throw ParameterException(INCORRECT_RECIPIENT)
        }

    }

    fun getFirstIndex(request: HttpServletRequest): Int {
        val firstIndex: Int
        try {
            firstIndex = Integer.parseInt(request.getParameter(FIRST_INDEX_PARAMETER))
            if (firstIndex < 0) {
                return 0
            }
        } catch (e: NumberFormatException) {
            return 0
        }

        return firstIndex
    }

    fun getLastIndex(request: HttpServletRequest): Int {
        val lastIndex: Int
        try {
            lastIndex = Integer.parseInt(request.getParameter(LAST_INDEX_PARAMETER))
            if (lastIndex < 0) {
                return Integer.MAX_VALUE
            }
        } catch (e: NumberFormatException) {
            return Integer.MAX_VALUE
        }

        return lastIndex
    }

    fun getCreationBytes(request: HttpServletRequest): ByteArray? {
        try {
            return request.getParameter(CREATION_BYTES_PARAMETER).parseHexString()
        } catch (e: RuntimeException) {
            throw ParameterException(INCORRECT_CREATION_BYTES)
        }


    }

    fun getATLong(request: HttpServletRequest): String {
        val hex = request.getParameter(HEX_STRING_PARAMETER)
        val bf = ByteBuffer.allocate(8)
        bf.order(ByteOrder.LITTLE_ENDIAN)
        bf.put(hex.parseHexString())
        return bf.getLong(0).toUnsignedString()
    }

    fun getAmountNQT(request: HttpServletRequest): Long {
        val amountValueNQT = request.getParameter(AMOUNT_NQT_PARAMETER).emptyToNull()
                ?: throw ParameterException(MISSING_AMOUNT)
        val amountNQT: Long
        try {
            amountNQT = java.lang.Long.parseLong(amountValueNQT)
        } catch (e: RuntimeException) {
            throw ParameterException(INCORRECT_AMOUNT)
        }

        if (amountNQT <= 0 || amountNQT >= Constants.MAX_BALANCE_NQT) {
            throw ParameterException(INCORRECT_AMOUNT)
        }
        return amountNQT
    }
}
